lib/commands/recordscreen.js

```diff
@@ -1,50 +1,209 @@
+import _ from 'lodash';
+import _fs from 'fs';
+import url from 'url';
 import { retryInterval } from 'asyncbox';
 import B from 'bluebird';
-import { util } from 'appium-support';
+import { util, fs } from 'appium-support';
 import log from '../logger';
+import temp from 'temp';
+import request from 'request-promise';
+import Ftp from 'jsftp';
 
 
 let commands = {}, extensions = {};
 
 const RETRY_PAUSE = 1000;
+const MAX_RECORDING_TIME_SEC = 60 * 3;
+const SCREENRECORD_BINARY = 'screenrecord';
+const DEFAULT_EXT = '.mp4';
+
+async function extractCurrentRecordingPath (adb, pids) {
+  let lsofOutput = '';
+  try {
+    const {output} = await adb.shell(['lsof', '-p', pids.join(',')]);
+    lsofOutput = output;
+  } catch (err) {
+    log.warn(`Cannot extract the path to the current screen capture. ` +
+             `Original error: ${err.message}`);
+    return null;
+  }
+  log.debug(`Got the following output from lsof: ${lsofOutput}`);
+  const pattern = new RegExp(/\d+\s+(\/.*\.mp4)/);
+  const matches = pattern.exec(lsofOutput);
+  return _.isEmpty(matches) ? null : _.last(matches);
+}
+
+async function uploadMediaToHttp (localFileStream, remoteUrl, uploadOptions = {}) {
+  const {user, pass, method} = uploadOptions;
+  const options = {
+    url: remoteUrl.href,
+    method: method || 'PUT',
+    multipart: [{ body: localFileStream }],
+  };
+  if (user && pass) {
+    options.auth = {user, pass};
+  }
+  log.debug(`Http upload options: ${JSON.stringify(options)}`);
+
+  const response = await request(options);
+  const responseDebugMsg = `Response code: ${response.statusCode}. Response body: ${JSON.stringify(response.body)}`;
+  log.debug(responseDebugMsg);
+  if (response.statusCode >= 400) {
+    throw new Error(`Cannot upload the recorded media to '${remoteUrl.href}'. ${responseDebugMsg}`);
+  }
+}
+
+async function uploadMediaToFtp (localFileStream, remoteUrl, uploadOptions = {}) {
+  const {user, pass} = uploadOptions;
+  const options = {
+    host: remoteUrl.hostname,
+    port: remoteUrl.port || 21,
+  };
+  if (user && pass) {
+    options.user = user;
+    options.pass = pass;
+  }
+  log.debug(`FTP upload options: ${JSON.stringify(options)}`);
+
+  return await new B((resolve, reject) => {
+    new Ftp(options).put(localFileStream, remoteUrl.pathname, (err) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve();
+      }
+    });
+  });
+}
+
+function toReadableSizeString (bytes) {
+  if (bytes >= 1048576) {
+    return `${parseFloat(bytes / 1048576.0).toFixed(2)} MB`;
+  } else if (bytes >= 1024) {
+    return `${parseFloat(bytes / 1024.0).toFixed(2)} KB`;
+  }
+  return `${bytes} B`;
+}
+
+async function uploadRecordedMedia (adb, pathOnDevice, remotePath = null, uploadOptions = {}) {
```

Comment:
all these helper functions can be also reused in iOS. Perhaps, we could move them to some common lib. the only question is to what lib

Reply:
The only place would be `appium-support`.